sw/device/lib/crypto/drivers/hmac.c

```diff
@@ -8,148 +8,420 @@
-                   UINT32_MAX);
+static void wait_hmac_done(void) {
+  uint32_t intr_reg = 0;
+  while (bitfield_bit32_read(intr_reg, HMAC_INTR_STATE_HMAC_DONE_BIT) == 0) {
+    intr_reg = abs_mmio_read32(kHmacBaseAddr + HMAC_INTR_STATE_REG_OFFSET);
+  }
+
+  // Clear the interrupt by writing 1, because `INTR_STATE` is rw1c type.
+  abs_mmio_write32(kHmacBaseAddr + HMAC_INTR_STATE_REG_OFFSET, intr_reg);
 }
 
 /**
- * Initialize the HMAC block.
+ * Clear the state of HMAC HWIP so that further driver calls can use it.
  *
- * If `enable_hmac` is `kHardenedBoolTrue`, starts the block in HMAC mode. If
- * it is `kHardenedBoolFalse`, starts in SHA256 mode.
+ * This function cannot force stop HWIP, and ongoing operations will not simply
+ * stop by deasserting `sha_en` bit. Instead it should be used after HWIP
+ * raises `hmac_done` interrupt (see `wait_hmac_done` function).
  *
- * @param enable_hmac Whether to start HMAC mode.
- * @param big_endian Whether to return digest in big-endian form.
+ * It also clears the internal state of HWIP by overwriting sensitive values
+ * with 1s.
  */
-static void hmac_init(hardened_bool_t enable_hmac) {
-  uint32_t cfg = 0;
-  // Digest should be big-endian to match the SHA-256 specification.
-  cfg = bitfield_bit32_write(cfg, HMAC_CFG_DIGEST_SWAP_BIT, true);
-  // Message should be little-endian to match Ibex's endianness.
-  cfg = bitfield_bit32_write(cfg, HMAC_CFG_ENDIAN_SWAP_BIT, false);
-  cfg = bitfield_bit32_write(cfg, HMAC_CFG_SHA_EN_BIT, true);
-  if (launder32(enable_hmac) == kHardenedBoolTrue) {
-    HARDENED_CHECK_EQ(enable_hmac, kHardenedBoolTrue);
-    cfg = bitfield_bit32_write(cfg, HMAC_CFG_HMAC_EN_BIT, true);
-  } else {
-    HARDENED_CHECK_EQ(enable_hmac, kHardenedBoolFalse);
-    cfg = bitfield_bit32_write(cfg, HMAC_CFG_HMAC_EN_BIT, false);
+static void hmac_hwip_clear(void) {
+  // Do not clear the config yet, we just need to deassert sha_en, see #23014.
+  // TODO handle digest size changes.
+  uint32_t cfg_reg = abs_mmio_read32(kHmacBaseAddr + HMAC_CFG_REG_OFFSET);
+  cfg_reg = bitfield_bit32_write(cfg_reg, HMAC_CFG_SHA_EN_BIT, false);
+  abs_mmio_write32(kHmacBaseAddr + HMAC_CFG_REG_OFFSET, cfg_reg);
+
+  // TODO(#23191): Use a random value from EDN to wipe.
+  abs_mmio_write32(kHmacBaseAddr + HMAC_WIPE_SECRET_REG_OFFSET, UINT32_MAX);
+}
+
+/**
+ * Restore the internal state of HWIP from `ctx` struct, and resume the
+ * operation.
+ *
+ * The first HWIP operation requires the call of `start` instead of `continue`.
+ * Therefore, `ctx->hw_started` flag is used to distinguish the first call. This
+ * function also updated `ctx->hw_started` after the first such call.
+ *
+ * If this function is being called from `ctx` object with previously stored
+ * context (i.e. `ctx->hw_started = true`), then this state is restored.
+ *
+ * @param ctx Context from which values are written to CSRs.
+ */
+static void restore_context(hmac_ctx_t *ctx) {
```

Comment:
The ROM and cryptolib have been trying to maintain a naming convention where the verb is used as the function suffix. For example:

```c
context_save()
context_restore()
msf_fifo_write()
etc
```